ask user to terminate or restart a task if it is active

elaihau opened the pull request:
With changes in this pull request, Theia offers users the flexibility of terminating or restarting a task, if the user tries to run a task that is actively running.
This pull request resolves https://github.com/eclipse-theia/theia/issues/6618#issuecomment-558047443, and resolves https://github.com/eclipse-theia/theia/issues/5269

Signed-off-by: Liang Huang <liang.huang@ericsson.com>

#### How to test

1. Start a long running task, or a background task
2. Before the task that is started in step 1 finishes, run it for one more time
    - Expected: Users should see a prompt. From the prompt they are offered 2 options: terminate the task (started in step 1), or restart the task.


![Peek 2019-11-30 19-02](https://user-images.githubusercontent.com/37082801/69908008-e7a7ea00-13ae-11ea-8738-22fa208dd3b6.gif)



#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

akosyakov commented:
@elaihau Is it aligned with VS Code UX?

RomanNikitenko commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +409,40 @@ export class TaskService implements TaskConfigurationClient {
+            }
+            this.messageService.info(`The task \'${taskName}\' is already active`, 'Terminate Task', 'Restart Task').then(actionTask => {
+                if (actionTask) {
+                    this.taskServer.kill(taskId);
```
Maybe it makes sense to use [the corresponding method](https://github.com/eclipse-theia/theia/blob/c4b4d3995ba81a80e36cd62d290dce2b23c8fd82/packages/task/src/browser/task-service.ts#L683) for terminating a task, this one handles errors

elaihau commented:
> @elaihau Is it aligned with VS Code UX? i.e. if task is finished and then started again the same terminal will be used
> 
> To resolve #5269 we should print to the terminal like in [#5269 (comment)](https://github.com/eclipse-theia/theia/issues/5269#issuecomment-496800867). Would it happen with this PR?

No what you described here is not implemented in this PR. 

We have a separated GH issue for it https://github.com/eclipse-theia/theia/issues/5332

And the features of 
- reusing the terminal
- printing the message(s) in the terminal
are controled by the `presentation` config (https://code.visualstudio.com/docs/editor/tasks#_output-behavior) which is a larger problem. I will resolve it bit by bit.

elaihau commented:
> @elaihau
> I tested your changes and noticed that the difference with VS Code behavior is:
> 
> * VS Code hides the bottom panel when user selects `Terminate Task`

@RomanNikitenko thank you for the review ! I noticed that too :)
It is not  implemented in the initial version of this patch because I am wondering if it is truly user friendly. For myself, as a developer, the terminal output is interesting even if the task is terminated. I wish the terminal stays visible after hitting the `Terminate Task` button, rather than seeing it varnish all of sudden. 

But that's just my opinion ... I should follow the vscode behavior. And if users are unhappy someone can file a GH issue :)  

> * another terminal widget is used if a task is completed and then started again (the current PR behavior)


Yes I can change my PR to follow this vscode behaviors.

RomanNikitenko commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +409,40 @@ export class TaskService implements TaskConfigurationClient {
+        }
+    }
+
+    private async doRunTask(task: TaskConfiguration, option?: RunTaskOption): Promise<TaskInfo | undefined> {
```
Can we have `protected` modifier here?
I think it will be more flexible for cases when some class extends `TaskService`
From this point of view also would nice to have at least `protected` `restartTask` method.
For example, at the moment `attach` method is `public` and it's possible to override this logic, provide own logic. 
`restartTask` logic is relative to it.
wdyt?

lmcbout commented:
Comparing VSCode and Theia
I like in VSCode the task command being display at the beginning of the execution in the terminal.
When stopping the task, VSCode remove the terminal while Theia keeps it, prefer the Theia version.
 
Suggestion: Would it be nice to put the message  "... terminated by signal SIGHUP" at the end of the trace in the terminal and/or popping up a dialog to show the warning. In this case, the end-user would see in the terminal he/she stops the task while keeping the terminal view available!

lmcbout commented:
I just noticed something. when you select the "Terminate task", the terminal stays OK
But if you resize the terminal, look at the server side console, you receive a lot of the following: "root ERROR Couldn't resize terminal 41, because it doesn't exist."

vince-fugnitto commented:
> I just noticed something. when you select the "Terminate task", the terminal stays OK
> But if you resize the terminal, look at the server side console, you receive a lot of the following: "root ERROR Couldn't resize terminal 41, because it doesn't exist."

I don't see the warning following your procedure.
In any case, I don't think it is harmful, the issue is not caused by this PR: https://github.com/eclipse-theia/theia/issues/4229

vince-fugnitto commented:
> Comparing VSCode and Theia
> I like in VSCode the task command being display at the beginning of the execution in the terminal.
> When stopping the task, VSCode remove the terminal while Theia keeps it, prefer the Theia version.
> 
> Suggestion: Would it be nice to put the message "... terminated by signal SIGHUP" at the end of the trace in the terminal and/or popping up a dialog to show the warning. In this case, the end-user would see in the terminal he/she stops the task while keeping the terminal view available!

@lmcbout I'd argue it's an enhancement to the current behavior. When we add support to display information regarding the task in the terminal https://github.com/eclipse-theia/theia/issues/5332 we can also support displaying the exit message.

elaihau commented:
> Comparing VSCode and Theia
> I like in VSCode the task command being display at the beginning of the execution in the terminal.
> When stopping the task, VSCode remove the terminal while Theia keeps it, prefer the Theia version.

> Suggestion: Would it be nice to put the message "... terminated by signal SIGHUP" at the end of the trace in the terminal and/or popping up a dialog to show the warning. In this case, the end-user would see in the terminal he/she stops the task while keeping the terminal view available!

@lmcbout Thank you for the feedback
Yes I love seeing that message too :) And the feature (i.e., adding messages before and after running the task) you described here is controled by the presentation config (https://code.visualstudio.com/docs/editor/tasks#_output-behavior) which is a larger problem. I will resolve it bit by bit.

elaihau commented:
@RomanNikitenko I resolved your comments and patched the code.

akosyakov commented:
@elaihau thanks for the explanation!

> are controled by the presentation config (https://code.visualstudio.com/docs/editor/tasks#_output-behavior) which is a larger problem. I will resolve it bit by bit.

Please make sure to create issues for all missing features to make progress visible for others.

RomanNikitenko commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +410,46 @@ export class TaskService implements TaskConfigurationClient {
+                        const parent = terminal.parent;
+                        terminal.close();
+                        if (actionTask === 'Terminate Task' && parent && parent.id === BOTTOM_AREA_ID) {
+                            parent.hide();
```
The behavior can be improved for case when user has a few running tasks.
So if user terminates one task, but another task is running - output for second task should be displayed, we shouldn't hide the bottom area.
I don't mind to improve the behavior in another PR, up to you.

RomanNikitenko commented:
> @RomanNikitenko I resolved your comments and patched the code.

@elaihau 
thank you very much!
wdyt about separate method `restartTask` as I mentioned [here](https://github.com/eclipse-theia/theia/pull/6668#discussion_r352580540)

lmcbout commented:
Just tested the latest commit dc759550de6bc26eb5db3a4508d01f9664a7ebd8 
If I open a terminal at the bottom, then run a long running task + terminate the task, when terminating the task, the terminal window is also removed from the bottom view area with the terminated task terminal view.

![BottomTerminalRemoved](https://user-images.githubusercontent.com/419051/70050301-9c532e80-159c-11ea-9616-10b94873f9e9.gif)

akosyakov commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +410,46 @@ export class TaskService implements TaskConfigurationClient {
+                if (actionTask) {
+                    await this.kill(taskId);
+                    if (terminal) {
+                        const parent = terminal.parent;
```
That looks fragile, terminals can be moved between areas, some product cat integrate terminal in another widget.

We should not assume anything about areas or structure of widgets here.

elaihau replied:
> That looks fragile, terminals can be moved between areas, some product cat integrate terminal in another widget.

Yes that's why I am checking `parent.id === BOTTOM_AREA_ID`. In my opinion, it makes sense to hide the parent only if it is in the bottom panel. What do you think?

akosyakov commented on packages/task/src/browser/task-service.ts:
```diff
@@ -648,7 +689,7 @@ export class TaskService implements TaskConfigurationClient {
             TERMINAL_WIDGET_FACTORY_ID,
             <TerminalWidgetFactoryOptions>{
                 created: new Date().toString(),
```
this code for long time bother me: `TerminalWidgetFactoryOptions` === terminal id, i.e. `TerminalWidgetFactoryOptions` should be a minimal amount of attributes uniquely identifying an instance, e.g. for and editor, it is a uri without a fragment.

I cannot get why `created` and `destroyTermOnClose` should be a part of a terminal id.

elaihau replied:
`created` and `destroyTermOnClose` are not part of a terminal id. As far as I can tell they are part of the `TerminalWidgetFactoryOptions`.

they were added long time ago in this commit 
```
  Sven Efftinge, 2 years ago   (September 17th, 2017 10:48am)

[storage] layout preservation and storage service implementation

Signed-off-by: Sven Efftinge <sven.efftinge@typefox.io>

c7c46de ≪ ≡ ↗ ⋯

Changes   —   75a48a9  ⟷  c7c46de
```
I will take a look at the commit and see what I can do in my future change

elaihau replied:
i looked into the commit where Sven introduce the interface: https://github.com/eclipse-theia/theia/commit/c7c46de3e6b4f4f9d1d94e8efb2081ed916a8afc#diff-bfac6f937a5be80c608b3b3ad2f865bdR31

My guess is that using the time stamp on the `created` property is probably  the easiest solution to identify terminals, because in Theia users are free to create as many terminal widget as they want. 

In my opinion `TerminalWidgetFactoryOptions` contains an ID but it is not used just as an ID.
```
interface TerminalWidgetFactoryOptions extends Partial<TerminalWidgetOptions> {
    created: string
}
```
As you can see in the line you added your comment, it is used as an object to pass "options" into the widget manager, so that the widget manager had sufficient information to create the terminal widget.

akosyakov replied:
ok

elaihau commented:
@RomanNikitenko I refactored the code to add `restartTask` and `terminateTask`.

Also I fixed the bug @lmcbout and @RomanNikitenko found. With the most recent change, the bottom panel wouldn't be hidden if there are other terminals open.

akosyakov commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +410,73 @@ export class TaskService implements TaskConfigurationClient {
+                    this.shell.activateWidget(terminal.id); // make the terminal visible and assign focus
+                }
+            }
+            this.messageService.info(`The task \'${taskName}\' is already active`, 'Terminate Task', 'Restart Task').then(async actionTask => {
```
Should not we use async/await (instead of `.then`) and return a task from `this.restartTask`?

akosyakov commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +410,73 @@ export class TaskService implements TaskConfigurationClient {
+                terminal.close();
+                const hasOtherTerminals = this.terminalService.all.length > 0;
+                if (!hasOtherTerminals && parent && parent.id === BOTTOM_AREA_ID) {
+                    parent.hide();
```
Do we need it at all? Should not we check at least that there is no other widgets in the bottom area interested for a user?

elaihau replied:
I agree with you Anton. Like I described [here](https://github.com/eclipse-theia/theia/pull/6668#issuecomment-561790788), I am complying with VS code behavior although I dislike it. 
I am not a product manager so my words of "I like it" or "i dislike it" don't count :)

elaihau replied:
in my opinion, 
- we should not close the terminal after the task is terminated. WHY vscode assumes the terminal output is not interesting to developers after terminating the task?
- we should NOT hide the panel, as there might be other widgets  in the panel that I want to keep watching. 

again, I have no idea why vs code does it this way. if most reviewers agree with what I proposed. I will update the code and make it work the Theia way. @akosyakov @RomanNikitenko @lmcbout @vince-fugnitto 

Please share your opinions and insight with me. Thank you !

elaihau replied:
Ooops I am not sure why my comment disappeared so I am going to post it again.

In my opinion, 
- the terminal should NOT be closed after terminating the task, because the task output could be interesting to the user, and
- the bottom panel should NOT be hidden, because there could be other widgets docked. 

If most reviewers here agree with this proposal, I will update my code to make the task termination the Theia way. 

Could you please share your insights ? @akosyakov @lmcbout @RomanNikitenko @vince-fugnitto

vince-fugnitto replied:
@elaihau I think you should go with your gut and keep the terminal and bottom area displayed :)

I think each of us (@lmcbout, yourself and I) all agree that VS Code's behavior is odd.

In my opinion:
- the bottom panel might have multiple different widgets present (ex: `memory view`), and hiding it when a task is incorrect. There is value in displaying the bottom area.
- the termination of a task may still hold valuable information and output that a user might care about
- I think the behavior of closing an entire area of the application is annoying. I think it should be up to the user to manually make that decision and not for the application to assume it should be closed.

RomanNikitenko replied:
> In my opinion,
> 
> * the terminal should NOT be closed after terminating the task, because the task output could be interesting to the user, and
> * the bottom panel should NOT be hidden, because there could be other widgets docked.
> 
> Again, I don't know WHY vscode closes the terminal and hides the panel. I dislike it.
> 
> If most reviewers here agree with this proposal, I will update my code to make the task termination the Theia way.
> 
> Could you please share your insights ? @akosyakov @lmcbout @RomanNikitenko @vince-fugnitto

I commented about the different behavior from VS Code, because of alignment theia behavior with VS Code behavior, but I'm totally agree with @elaihau and as a developer I prefer do **not** hide the bottom area and do **not** close terminal after terminating a task. User has opportunity to hide bottom panel using hotkey.

akosyakov replied:
let's remove it for now then and wait for real issues, it also makes code simpler

lmcbout replied:
I agree with @vince-fugnitto and @elaihau : The behavior to remove the terminal when task terminates is odd. I prefer keeping the terminal open when the task is terminated. One thing, we need to be careful because the terminal seems not be registered after, so if we resize the view after termination, I can see a trace in the back-end side, not harmful, but not necessary to clutter the console with the error either.  So my opinion would be to be able to see the terminal view even after task termination, not hide it.

akosyakov commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +410,73 @@ export class TaskService implements TaskConfigurationClient {
+        if (terminalId) {
+            const terminal = this.terminalService.getById(this.getTerminalWidgetId(terminalId));
+            if (terminal) {
+                terminal.close();
```
does not it already happen in `terminateTask`?

lmcbout commented:
There are still an error with the bottom panel.
How to reproduce:
- Start a long running process
- Open the Problem View
- Terminate the long running task
--> The bottom panel will go away closing the problem view
Using commit 4d8dfd7028619044d30248b100af39d454c23197

![TerminalTaskClosePanel](https://user-images.githubusercontent.com/419051/70142056-6b3f3080-1666-11ea-8d09-76ef24da3210.gif)

lmcbout commented:
Question:
When the long running task is moved to another panel, let say the right| left panel, and you restart the long running task, where should it be shown on he restart: Right, left or bottom panel?
Currently, it re-open on the bottom panel.

vince-fugnitto commented:
> Question:
> When the long running task is moved to another panel, let say the right| left panel, and you restart the long running task, where should it be shown on he restart: Right, left or bottom panel?
> Currently, it re-open on the bottom panel.

In order to support such a feature, we'd have to extend the `TerminalWidgetOptions` to specify where to spawn the terminal widget. At the moment, terminals are always created in the `bottom area`:

https://github.com/eclipse-theia/theia/blob/d5c81105393a811de3097ff06bdf64ab98ef3d80/packages/core/src/browser/shell/application-shell.ts#L1605-L1625

elaihau commented:
> * Terminate the long running task
>   --> The bottom panel will go away closing the problem view
>   Using commit [4d8dfd7](https://github.com/eclipse-theia/theia/commit/4d8dfd7028619044d30248b100af39d454c23197)

@lmcbout Yes and vs code works this way too. I dislike it but my plan is comply with the vs code behavior - if anyone hates it s/he could create a ticket 

![Peek 2019-12-04 13-59](https://user-images.githubusercontent.com/37082801/70172327-73b35d80-169e-11ea-9f39-2ccaa14a80e2.gif)

elaihau commented:
> Question:
> When the long running task is moved to another panel, let say the right| left panel, and you restart the long running task, where should it be shown on he restart: Right, left or bottom panel?
> Currently, it re-open on the bottom panel.

Vincent's correct. Theia always spawn the terminal at the bottom panel. And we will need change the interface to include an option that controls the area that terminal is expected to be spawned.

akosyakov commented:
> Question:
When the long running task is moved to another panel, let say the right| left panel, and you restart the long running task, where should it be shown on he restart: Right, left or bottom panel?
Currently, it re-open on the bottom panel.

I think it should be another issue. We can ignore it for now. Don't think a user will move it.

elaihau commented:
In the latest patch I removed the code that closes the terminal & hides the panel. 

Also, Anton's  review comments are addressed.

Thank you all the for the feedback !

akosyakov commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +409,52 @@ export class TaskService implements TaskConfigurationClient {
+            if (selectedAction === 'Terminate Task') {
+                this.terminateTask(matchedRunningTaskInfo);
+            } else if (selectedAction === 'Restart Task') {
+                this.restartTask(matchedRunningTaskInfo, option);
```
Should not we add `return` here, if some code is waiting it should receive a proper task info not?

akosyakov commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +409,52 @@ export class TaskService implements TaskConfigurationClient {
+     */
+    protected async terminateTask(activeTaskInfo: TaskInfo): Promise<void> {
+        const taskId = activeTaskInfo.taskId;
+        this.kill(taskId);
```
`await` is missing?

akosyakov commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +409,52 @@ export class TaskService implements TaskConfigurationClient {
+            }
+            const selectedAction = await this.messageService.info(`The task \'${taskName}\' is already active`, 'Terminate Task', 'Restart Task');
+            if (selectedAction === 'Terminate Task') {
+                this.terminateTask(matchedRunningTaskInfo);
```
`await` is missing? 

Should we also add `return undefined` here after awaiting?

vince-fugnitto commented on packages/task/src/browser/task-service.ts:
```diff
@@ -166,7 +166,7 @@ export class TaskService implements TaskConfigurationClient {
                 taskIdentifier = this.taskNameResolver.resolve(task);
             }
-            this.messageService.info(`Task ${taskIdentifier} has been started`);
+            this.messageService.info(`Task \'${taskIdentifier}\' has been started`);
```
[minor] add punctuation to align with other messages?

https://github.com/eclipse-theia/theia/pull/6668/files#diff-794cdf6ef93aef7cc827f31f92ebe226R220
https://github.com/eclipse-theia/theia/pull/6668/files#diff-794cdf6ef93aef7cc827f31f92ebe226R227

vince-fugnitto commented on packages/task/src/browser/task-service.ts:
```diff
@@ -409,6 +409,52 @@ export class TaskService implements TaskConfigurationClient {
+                    this.shell.activateWidget(terminal.id); // make the terminal visible and assign focus
+                }
+            }
+            const selectedAction = await this.messageService.info(`The task \'${taskName}\' is already active`, 'Terminate Task', 'Restart Task');
```
[minor] do we want to align further and display the workspace folder name?

```
The task 'long running test task (test-resources)' is already active.
```

elaihau commented:
Thank you for the review Anton and Vincent ! Comments are resolved.

vince-fugnitto commented on packages/task/src/browser/task-service.ts:
```diff
@@ -210,27 +211,29 @@ export class TaskService implements TaskConfigurationClient {
+    private getTaskIdentifier(taskConfig: TaskConfiguration): string {
+        const taskName = this.taskNameResolver.resolve(taskConfig);
+        const sourceStrUri = this.taskSourceResolver.resolve(taskConfig);
+        return `${taskName} (${new URI(sourceStrUri).displayName})`;
```
[minor] do we need to use `labelProvider.getName(uri)` for such a case?:
https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#use-short-name

RomanNikitenko commented:
I tested the changes after last update and I can see the following behavior for the `long running task`:

![restart_new_behavior](https://user-images.githubusercontent.com/5676062/70358684-930ae000-1882-11ea-8d6b-06fe7258b583.gif)

For `Terminate Task` action:
- the task is terminated
- the terminal widget is open after terminating the task  

For `Restart Task` action:
- the task is terminated
- the terminal widget is open after terminating the task
- the task is started
- the new terminal is used for the output   

May be it makes sense to replace the gif in the PR description by gif which reflects the actual behavior. 

Also I tested for `detected` `tsc` tasks and it works well for me except for the one case:
https://youtu.be/Ka0NCENTqw4

So I run `detected` task for `che-theia` project, but the notification and the `recently used` section tell me that task for `theia` project is running.
The attempt to run the task for `che-theia` project leads to displaying the prompt `terminate` or `restart` the task.

I believe that the problem is **not** related to the current PR changes, something wrong with running `detected` tasks in multi-root workspaces case. So it can be reproduced for very specific case:
- multi-root workspace
- the tasks have the similar configuration and have only different `scope` (I guess, we need investigate it)  

@elaihau 
could you confirm that you can reproduce it and I'll create the separate issue for this bug.

elaihau commented:
@RomanNikitenko Yes I can reproduce the bug you described with NPM extension. 

There is something going wrong with the way that we search tasks from the cached data.

elaihau commented:
@RomanNikitenko I created https://github.com/eclipse-theia/theia/issues/6715 and will merge this one. Thanks !
